BD-3377 Triggered IAMs in Canvas EA (#7938)
braze-inc/braze-docs: _docs/_hidden/private_betas/canvas_triggered_iams.md

---
article_title: In-App Messages in Canvas
permalink: /canvas_triggered_in-app_messages/
page_type: reference
description: "This reference article describes features and nuances specific to Canvas in-app messages, which you can add to your Canvas to show rich messaging."
---

# In-app messages in Canvas

> In-app messages can be added as part of your Canvas journey to show rich messaging when your customer engages with your app. This article describes features and nuances specific to Canvas in-app messages.

Before continuing, you should have already [created your Canvas]({{site.baseurl}}/user_guide/engagement_tools/canvas/create_a_canvas/create_a_canvas/) and set up delay and audience options.

In the Canvas builder, add a [Message]({{site.baseurl}}/user_guide/engagement_tools/canvas/canvas_components/message_step/) step and select **In-App Message** as your **Messaging Channel**. You can customize [when your message will expire](#in-app-message-expiration) and which [advancement behavior](#advancement-behavior) it will have.

## Triggered in-app messages

> After any delays pass and the audience options are checked, in-app messages are set live when a user reaches the Message step. If a user starts a session and performs the trigger event for the in-app message, the user will see the in-app message. 

Reviewer: ```suggestion
After any delays pass and the audience options are checked, in-app messages are set to live when a user reaches the Message step. If a user starts a session and performs the trigger event for the in-app message, the user will see the in-app message. 
```

> For Canvas steps that have action-triggered entry, users can enter the Canvas mid-session. In-app messages aren't set live until a session starts, so if a user is in the middle of the session when they reach the Message step, they won't receive the in-app message until they start another session and perform the relevant trigger.

Reviewer: ```suggestion
For Canvas steps that have action-triggered entry, users can enter the Canvas mid-session. In-app messages aren't set to live until a session starts, so if a user is in the middle of the session when they reach the Message step, they won't receive the in-app message until they start another session and perform the relevant trigger.
```

![]({% image_buster /assets/img_archive/canvas_expiration2.png %}){: style="max-width:85%"}

| Option | Description | Example |
|---|---|---|
| **A duration after the step is available** | Sets the in-app message to expire relative to when the step becomes available to the user. | An in-app message with a two-day expiration would become available after the step's delay elapses and audience options are checked. It would then be available for 2 days (48 hours) and during those two days, users may see the in-app message if they open the app. |
| **On a specific date and time** | Select a specific date and time when the in-app message will be no longer available. | If you have a sale that ends on November 30, 2024, select this option so that users no longer see the associated in-app message when the sale ends. |
{: .reset-td-br-1 .reset-td-br-2 .reset-td-br-3}

### Use cases

When should you use this feature? Braze highly recommends that you consider using this feature in your promotional and onboarding Canvases.

{% tabs %}
  {% tab Promotional %}

> Promotions, coupons, and sales often have hard expiration dates. The following Canvas should alert your users at the most opportune times that there is a promotion they may use, and perhaps influence a purchase. This promotion expires on February 28, 2019, at 11:15 am in the company's time zone.

Reviewer: ```suggestion
Promotions, coupons, and sales often have hard expiration dates. The following Canvas should alert your users at the most opportune times that there is a promotion they may use, and perhaps influence a purchase. This promotion expires on February 28, 2019, at 11:15 am in your company's time zone.
```

  {% endtab %}
{% endtabs %}

## Prioritizing in-app messages

A user can trigger two in-app messages within your Canvas at the same time. When this happens, Braze will adhere to the following priority order to determine which in-app message is displayed. 

Select **Set exact priority** and drag different Canvas steps to reorder their priority for the Canvas. By default, steps earlier in a Canvas variant will display before later steps. After your steps are in your preferred order of prioritization, select **Apply sort**.

![]({% image_buster /assets/img_archive/canvas_priority2.png %}){: style="max-width:85%"}

### Making changes to drafts of active Canvases

If you make changes to the in-app message priority in **Send Settings** of a draft of an active Canvas, these changes are applied directly to the active Canvas when the priority sorter is closed. However, in a Message step, the priority sorter wil be updated when the draft is launched since Canvas step settings apply at a step level. 

## Advancement behavior

> Message components automatically advance all users who enter the step. Note it doesn't wait for the in-app message to trigger or display. There is no requirement to specify message advancement behavior, making configuring the overall step simpler. If you want to implement the **Advance when message sent** option, add a separate [audience path]({{site.baseurl}}/user_guide/engagement_tools/canvas/canvas_components/audience_paths/) to filter users that didn't receive the previous step.

Reviewer: ```suggestion
Message components automatically advance all users who enter the step. Note that it doesn't wait for the in-app message to trigger or display. There is no requirement to specify message advancement behavior, making configuring the overall step simpler. If you want to implement the **Advance when message sent** option, add a separate [audience path]({{site.baseurl}}/user_guide/engagement_tools/canvas/canvas_components/audience_paths/) to filter users that didn't receive the previous step.
```

{% details Original Canvas editor %}

{% alert important %}
As of February 28, 2023, you can no longer create or duplicate Canvases using the original editor. This section is available for reference when understanding how advancement behavior works for steps with in-app messages.
{% endalert %}

> Canvases created in the original editor need to specify an advancement behavior—the criteria for advancement through your Canvas component. [Steps with only in-app messages](#steps-iam-only) have different advancement options than [steps with multiple message types](#steps-multiple-channels) (push, email, etc.). For in-app messages in a Canvas Flow workflow, this option is set to always immediately advance the audience.

Reviewer: ```suggestion
Canvases created in the original editor need to specify an advancement behavior—the criteria for advancement through your Canvas component. [Steps with only in-app messages](#steps-iam-only) have different advancement options than [steps with multiple message types](#steps-multiple-channels) (such as push or email). For in-app messages in a Canvas Flow workflow, this option is set to always immediately advance the audience.
```

Action-based delivery is not available for Canvas steps with in-app messages. Canvas steps with in-app messages must be scheduled. Instead, Canvas in-app messages will appear the first time that your user opens the app (triggered by the start session) after the scheduled message in the Canvas component has been sent to them.

If you have multiple in-app messages within one Canvas, a user must start multiple sessions to receive each of those individual messages.

> ![][2]

Reviewer: The images aren't displaying. I think you need to put the image paths here instead of using references 🤔

---
article_title: In-App Messages in Canvas
permalink: /canvas_triggered_in-app_messages/
page_type: reference
description: "This reference article describes features and nuances specific to Canvas in-app messages, which you can add to your Canvas to show rich messaging."
---

# In-app messages in Canvas

> In-app messages can be added as part of your Canvas journey to show rich messaging when your customer engages with your app. This article describes features and nuances specific to Canvas in-app messages.

Before continuing, you should have already [created your Canvas]({{site.baseurl}}/user_guide/engagement_tools/canvas/create_a_canvas/create_a_canvas/) and set up delay and audience options.

Add a [Message]({{site.baseurl}}/user_guide/engagement_tools/canvas/canvas_components/message_step/) step and choose **In-App Message** for your **Messaging Channel**. 

After any delays pass and the audience options are checked, the in-app message will be set live and users will see it if they open the app. In-app messages in Canvas may only be triggered by the `start session` trigger event—they can't be triggered by custom events in a Canvas component.

> For Canvas steps that have action-triggered entry, users can enter the Canvas mid-session. However, as noted above, in-app messages won't trigger until the next session starts, so these users would miss the initial in-app message since they weren't eligible to enter the Canvas before the session started.

Reviewer 2: This references full steps, not message steps. And the second page seems to reference Canvas instead of Canvas steps?

I'd expect it to say something like:

- In-app messages are set live when a user reaches the message step. If the user starts a session and performs the trigger event for the IAM, the user will see the IAM.
- In-app messages aren't set live until a session starts, so if a user is in the middle of the session when they reach the message step, they won't receive the IAM until they start another session and perform the relevant trigger.

    <td>Push</td>
    <td>N/A</td>
    <td>Message Sent</td>
    <td>After users have received their in-app message, they will receive a follow-up push to continue their onboarding.</td>
  </tr>
</tbody>
</table>

These push messages are spaced around an in-app message to make sure the user has visited the app and started their onboarding. This helps prevent any spam or out-of-order messaging that could dissuade users from visiting your app, and instead create a flowing, sensible order to their initial experiences with your app.

  {% endtab %}
{% endtabs %}

## Prioritizing in-app messages

A user can trigger two in-app messages within your Canvas at the same time. When this happens, Braze will adhere to the following priority order to determine which in-app message is displayed. 

Select **Set exact priority** and drag different Canvas steps to reorder their priority for the Canvas. By default, steps earlier in a Canvas variant will display before later steps. After your steps are in your preferred order of prioritization, select **Apply sort**.

![]({% image_buster /assets/img_archive/canvas_priority2.png %}){: style="max-width:85%"}

### Drafts of an active Canvas

> When editing a draft of an active Canvas, changes to the in-app message priority within **Send Settings** are not saved with a draft. These changes are applied directly to the active Canvas when the priority sorter modal is closed. However, in a Message step, the priority sorter will be updated when a user launches the draft since step settings apply at a step level.

Reviewer 2: This bit is accurate but a little confusing, could we make it easier to read?

## Advancement behavior

### Canvas Flow

> Message components automatically advance all users who enter the step. There is no requirement to specify message advancement behavior, making configuring the overall step simpler. If you want to implement the **Advance when message sent** option, add a separate [audience path]({{site.baseurl}}/user_guide/engagement_tools/canvas/canvas_components/audience_paths/) to filter users that didn't receive the previous step.

Reviewer 2: Could we clarify the first sentence to be **Message components automatically advance all users who enter the step. Note it doesn't wait for the IAM to trigger or display** or something like that
